lib/core/shared/store.js

```diff
@@ -0,0 +1,123 @@
+ * Once instantiated, this class can only access the index passed in the
+ * constructor
+ */
+class Store {
+  /**
+   * @param {Kuzzle} kuzzle
+   * @param {String} index
+   * @param {storeScopeEnum} scope
+   */
+  constructor (kuzzle, index, scope) {
+    this.kuzzle = kuzzle;
+    this.index = index;
+    this.scope = scope;
+
+    const methodsMapping = {
+      count:              `core:storage:${scope}:document:count`,
+      create:             `core:storage:${scope}:document:create`,
+      createCollection:   `core:storage:${scope}:collection:create`,
+      createOrReplace:    `core:storage:${scope}:document:createOrReplace`,
+      delete:             `core:storage:${scope}:document:delete`,
+      deleteByQuery:      `core:storage:${scope}:document:deleteByQuery`,
+      deleteCollection:   `core:storage:${scope}:collection:delete`,
+      deleteIndex:        `core:storage:${scope}:index:delete`,
+      exists:             `core:storage:${scope}:document:exist`,
+      get:                `core:storage:${scope}:document:get`,
+      getMapping:         `core:storage:${scope}:mappings:get`,
+      mExecute:           `core:storage:${scope}:document:mExecute`,
+      mGet:               `core:storage:${scope}:document:mGet`,
+      refreshCollection:  `core:storage:${scope}:collection:refresh`,
+      replace:            `core:storage:${scope}:document:replace`,
+      search:             `core:storage:${scope}:document:search`,
+      truncateCollection: `core:storage:${scope}:collection:truncate`,
+      update:             `core:storage:${scope}:document:update`,
+      updateByQuery:      `core:storage:${scope}:document:updateByQuery`,
+      updateCollection:   `core:storage:${scope}:collection:update`,
+      updateMapping:      `core:storage:${scope}:mappings:update`,
+    };
+
+    for (const [method, event] of Object.entries(methodsMapping)) {
+      this[method] = (...args) => this.kuzzle.ask(event, this.index, ...args);
+    }
+
+    // the scroll method is special: it doesn't need an index parameter
+    // we keep it for ease of use
+    this.scroll = (scrollId, opts) => this.kuzzle.ask(
+      `core:storage:${scope}:document:scroll`,
+      scrollId,
+      opts);
+  }
+
+  /**
+   * Initialize the index, and creates provided collections
+   *
+   * @param {Object} collections - List of collections with mappings to create
+   *
+   * @returns {Promise}
+   */
+  async init (collections = {}) {
+    const mutex = new Mutex(this.kuzzle, `Store.init(${this.index})`, {
+      timeout: -1,
+      ttl: 30000,
+    });
+
+    await mutex.lock();
+
+    try {
+      await this.createCollections(collections);
+    }
+    finally {
+      await mutex.unlock();
+    }
+  }
+
+  /**
+   * Creates collections with the provided mappings
+   *
+   * @param {Object} collections - collections with mappings
+   *
+   * @returns {Promise}
+   */
+  createCollections (collections) {
+    return Bluebird.map(
```

Comment:
You should limit concurrency on this one because if there is a lot of collections it will put too much pressure on ES. (Omniscient crashed because of something like this)